### PR TITLE
[feature][METEOR-727] select the z stream for the Localization z and cancel localization z

### PR DESCRIPTION
**Pull request by K4rishma**
…ization z

### CODE REVIEW

**Comment by K4rishma** [comment]
Create two separate commits after the review process : one for stream selection and another for allowing cancellation

File: src/odemis/acq/align/z_localization.py [hunk 1]
```diff
@@ -300,6 +315,30 @@ def _do_measure_z(f: Future, stigmator, angle: float, pos: Tuple[float, float],
...
+    except CancelledError:
+        raise
+
     finally:
-        f = stigmator.moveAbs({"rz": 0})
-        f.result(timeout=600)
+        f.running_subf = stigmator.moveAbs({"rz": 0})
+        f.running_subf.result(timeout=600)
+        with f._task_lock:
+            if f._task_state == CANCELLED:
```
Comment (nandishjpatel) [inline src/odemis/acq/align/z_localization.py:326]: Do you want the final sub-future task `stigmator.moveAbs({"rz": 0})` to handle cancel and raise an exception? or this final sub-future should always be executed?

File: src/odemis/acq/align/z_localization.py [hunk 2]
```diff
@@ -300,6 +315,30 @@ def _do_measure_z(f: Future, stigmator, angle: float, pos: Tuple[float, float],
...
+            f._task_state = FINISHED
+
+
+def _cancel_localization(future) -> bool:
+    """
+    Canceler of Z Localization.
+    :param future: the future that will be executing the task
+    :return: True if it successfully cancelled (stopped) the future
+    """
+    logging.debug("Canceling localization procedure...")
```
Comment (nandishjpatel) [inline src/odemis/acq/align/z_localization.py:337]:
```suggestion
    logging.debug("Cancelling localization procedure...")
```

**Review by pieleric — CHANGES_REQUESTED** [review]

File: src/odemis/gui/cont/acquisition/cryo_z_localization.py [hunk 3]
```diff
@@ -124,21 +131,65 @@ def _check_button_available(self, _):
...
+            else:
+                self._selected_stream = next(s for s in self._tab_data.streams.value if isinstance(s, FluoStream))
+
+    def _create_stream_menu(self, evt):
+        """Display active list of streams in the menu and check the selected stream when toggle button is clicked"""
+        menu = wx.Menu()
+        s = [s for s in self._tab_data.streams.value if isinstance(s, FluoStream)]
+        self._menu_to_stream = {}
+        # Keep the display order of streams in menu, same as,
+        # Display order in the "Streams" panel in the Localization Tab
```
[the remark this answers is not in the document]
Comment (K4rishma) [reply src/odemis/gui/cont/acquisition/cryo_z_localization.py:160]: That is correct, if the order of the stream panel streams and menu list streams matches if they are in the same order. I will rectify this.

**Review by pieleric — CHANGES_REQUESTED** [review]

**Review by pieleric — APPROVED** [review]

**Pull request merged**